src/docfx/lib/markdown/validation/ContentValidationExtension.cs

```diff
@@ -1,23 +1,39 @@
 // Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
 
+using System;
+using System.Collections.Immutable;
 using Markdig;
 using Microsoft.DocAsCode.MarkdigEngine.Extensions;
+using Microsoft.DocAsCode.MarkdigEngine.Validators;
 using Validations.DocFx.Adapter;
 
 namespace Microsoft.Docs.Build
 {
     internal static class ContentValidationExtension
     {
-        public static MarkdownPipelineBuilder UseContentValidation(this MarkdownPipelineBuilder builder, MarkdownContext markdownContext, string markdownValidationRulesPath)
-        {
-            var validatorProvider = new OnlineServiceMarkdownValidatorProvider(
-                new ContentValidationContext(markdownValidationRulesPath),
-                new ContentValidationLogger(markdownContext));
+        private static ImmutableArray<IMarkdownObjectValidator> _validators;
+        private static object locker = new object();
 
+        public static MarkdownPipelineBuilder UseContentValidation(this MarkdownPipelineBuilder builder, MarkdownContext markdownContext, Func<string> getMarkdownValidationRulesPath)
+        {
             return builder.Use(document =>
             {
-                foreach (var validator in validatorProvider.GetValidators())
+                if (_validators.IsDefaultOrEmpty)
+                {
+                    lock (locker)
```

Comment:
😱 😱 😱 😱 😱 😱 😱 can you fix the thread bug instead

Reply:
not the package bug...

Reply:
the locker is used to avoid repeated initialization...

Reply:
Duplicated initialization can be solved by using `Lazy<T>`